Non broken fetchcargo

symphorien opened the pull request:
###### Motivation for this change

This builds on the work of @erictapen https://github.com/NixOS/nixpkgs/pull/45889
In this comment https://github.com/NixOS/nixpkgs/pull/45889#issuecomment-418708320 @Mic92 suggested the output of cargo vendor (which is not guaranteed to be stable by upstream) be filtered/normalised to ensure we get a fixed output derivation, somewhat like fetchpatch does.
This is an attempt at implementing it.

What is normalized:
- order of key/value pairs
- spacing
- which quotes are used
- the vendoring directory name

what is not normalised
- the content of the toml (for example trailing / at the end or urls or which keys appear). Given the current state of cargo vendor, the set of keys which appear is deterministic https://github.com/alexcrichton/cargo-vendor/blob/master/src/main.rs#L48-L67 so I imagine this is fine.

fixes https://github.com/NixOS/nixpkgs/issues/41518

###### Things done

Tested that nix-du can build with `useRealVendorConfig = true` and that I can build the updated habitat 
from this PR https://github.com/NixOS/nixpkgs/pull/46138 without the cargo config hack they are forced to use.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"` (nothing changes)
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

erictapen committed:
buildRustPackage, fetchcargo: optionally use vendor config from cargo-vendor

By setting useRealVendorConfig explicitly to true, the actual (slightly
modified) config generated by cargo-vendor is used.

This solves a problem, where the static vendor config in
pkgs/build-support/rust/default.nix would not sufficiently replace all
crates Cargo is looking for.

As useRealVendorConfig (and writeVendorConfig in fetchcargo) default to
false, there should be no breakage in existing cargoSha256 hashes.

Nethertheless, imho using this new feature should become standard. A
possible deprecation path could be:

- introduce this patch
- set useRealVendorConfig explicitly to false whereever cargoSha256 is
  set but migration is not wanted yet.
- after some time, let writeVendorConfig default to true
- when useRealVendorConfig is true everywhere cargoSha256 is set and
  enough time is passed, `assert cargoVendorDir == null ->
  useRealVendorConfig;`, remove old behaviour
- after some time, remove all appearences of useRealVendorConfig and the
  parameter itself

symphorien committed:
fetchcargo: normalise cargo config to ensure determinism

Mic92 commented on doc/languages-frameworks/rust.section.md:
```diff
@@ -64,6 +65,14 @@ When the `Cargo.lock`, provided by upstream, is not in sync with the
 added in `cargoPatches` will also be prepended to the patches in `patches` at
 build-time.
 
+The `useRealVendorConfig` parameter tells cargo-vendor to include a Cargo
```
Do we really have an incompatibility compared to the current state? Is the output different for packages that build at the moment? I would like to enable this by default otherwise.

Mic92 replied:
Otherwise a warning or should be printed otherwise people will never know about this option or bother to migrate.

symphorien replied:
The incompatibility is: this changes the expected `cargoSha256`. So two approaches:
* we use the new scheme forcibly from now on and break every current hash
* or we deprecate the old behavior, let say one nixos release for people to switch to the new behavior, and then remove the old behavior.
This PR implements the second approach.
I agree about the warning.

Mic92 replied:
So you would not just add this:

 ```
  [source.crates-io]
  registry = 'https://github.com/rust-lang/crates.io-index'
  replace-with = 'vendored-sources'
  [source.vendored-sources]
  directory = '$(pwd)/$cargoDepsCopy'
```

to keep the checksum of existing packages in tact and then append the new content as required by the Cargo.toml?

Mic92 replied:
Or is there a problem with this approach?

symphorien replied:
the right approach is to *remove* the config when it only contains crates.io. But this is a great idea indeed. I just pushed a commit which should not break any cargoSha256. It is a mass rebuild though. (it rebuilds cargo)

symphorien committed:
fectchcargo: don't break old sha256

symphorien commented:
Here is a new commit which does not change any cargosha256 and is thus backward compatible.
I have tried to check that hashes actually don't change, but if reviewers can check it as well this would be nice.

bkchr reviewed (approved):
LGTM, especially that you save the generated `config`.

Mic92 committed:
fetchcargo: add type checking to cargo-vendor-normalise.py

Mic92 committed:
alacritty: switch back to upstream source

Thanks to https://github.com/NixOS/nixpkgs/pull/46362
We can now support git dependencies!

Mic92 commented on pkgs/build-support/rust/cargo-vendor-normalise.py:
```diff
@@ -0,0 +1,41 @@
+    return '"{}"'.format(escaped)
+
+
+def main() -> None:
```
I added type checking here and reformatted the code with `black` + some minor readability improvements.

Mic92 commented on pkgs/applications/misc/alacritty/default.nix:
```diff
@@ -51,18 +51,16 @@ let
-    fetchSubmodules = true;
+  version = "2018-08-05";
+
+  src = fetchFromGitHub {
```
I added this as a test, because it makes use of the new `.cargo/config`

Mic92 commented:
@GrahamcOfBorg build alacritty

Mic92 commented:
Can you rebase this on staging?

GrahamcOfBorg commented:
<!--REQUEST_ID=74048f67-0369-4a7a-b1af-999b121e008e-->
Unexpected error: command failed with exit code 1 on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46362&attempt_id=b202b564-1b46-4d5d-a70d-be094b206ce2)

Attempted: alacritty

<details><summary>Partial log (click to expand)</summary><p>

```
   Vendoring ws2_32-sys v0.2.1 (/private/tmp/nix-build-alacritty-unstable-2018-08-05-vendor.drv-0/source/cargo-home.4Ok/registry/src/github.com-1ecc6299db9ec823/ws2_32-sys-0.2.1) to /nix/store/7aw52wn8pcppggy92w5ga12gr46bgz4n-alacritty-unstable-2018-08-05-vendor/ws2_32-sys
   Vendoring x11-dl v2.18.3 (/private/tmp/nix-build-alacritty-unstable-2018-08-05-vendor.drv-0/source/cargo-home.4Ok/registry/src/github.com-1ecc6299db9ec823/x11-dl-2.18.3) to /nix/store/7aw52wn8pcppggy92w5ga12gr46bgz4n-alacritty-unstable-2018-08-05-vendor/x11-dl
   Vendoring xdg v2.1.0 (/private/tmp/nix-build-alacritty-unstable-2018-08-05-vendor.drv-0/source/cargo-home.4Ok/registry/src/github.com-1ecc6299db9ec823/xdg-2.1.0) to /nix/store/7aw52wn8pcppggy92w5ga12gr46bgz4n-alacritty-unstable-2018-08-05-vendor/xdg
   Vendoring xml-rs v0.7.0 (/private/tmp/nix-build-alacritty-unstable-2018-08-05-vendor.drv-0/source/cargo-home.4Ok/registry/src/github.com-1ecc6299db9ec823/xml-rs-0.7.0) to /nix/store/7aw52wn8pcppggy92w5ga12gr46bgz4n-alacritty-unstable-2018-08-05-vendor/xml-rs
   Vendoring yaml-rust v0.4.0 (/private/tmp/nix-build-alacritty-unstable-2018-08-05-vendor.drv-0/source/cargo-home.4Ok/registry/src/github.com-1ecc6299db9ec823/yaml-rust-0.4.0) to /nix/store/7aw52wn8pcppggy92w5ga12gr46bgz4n-alacritty-unstable-2018-08-05-vendor/yaml-rust
To use vendored sources, add this to your .cargo/config for this project:

builder for '/nix/store/7zqjmaaf0k5zl3yz5g9cc7z0x5vhki1j-alacritty-unstable-2018-08-05-vendor.drv' failed with exit code 2
cannot build derivation '/nix/store/x5y8kxq2cxbclphh113hd22g6kvamgmw-alacritty-unstable-2018-08-05.drv': 1 dependencies couldn't be built
error: build of '/nix/store/x5y8kxq2cxbclphh113hd22g6kvamgmw-alacritty-unstable-2018-08-05.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=74048f67-0369-4a7a-b1af-999b121e008e-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46362&attempt_id=e544843f-128c-4d12-b5e2-2c3944c5a2fb)

Attempted: alacritty

<details><summary>Partial log (click to expand)</summary><p>

```
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `u8`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0119`.
error: Could not compile `winit`.
warning: build failed, waiting for other jobs to finish...
error: build failed
builder for '/nix/store/69pgvahdf92rlg4qsbhifn98wnvzslhc-alacritty-unstable-2018-08-05.drv' failed with exit code 101
error: build of '/nix/store/69pgvahdf92rlg4qsbhifn98wnvzslhc-alacritty-unstable-2018-08-05.drv' failed
```
</p></details>

Mic92 commented:
@GrahamcOfBorg build fd

Mic92 commented:
It would be great if someone with macOS could look at the failure

GrahamcOfBorg commented:
<!--REQUEST_ID=74048f67-0369-4a7a-b1af-999b121e008e-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46362&attempt_id=bb1b94d9-baba-44fb-8de5-5057f9528d2f)

Attempted: alacritty

<details><summary>Partial log (click to expand)</summary><p>

```
post-installation fixup
gzipping man pages under /nix/store/04d6b7v15l2mg7aq24xsf9xzgjl18rj5-alacritty-unstable-2018-08-05/share/man/
strip is /nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/04d6b7v15l2mg7aq24xsf9xzgjl18rj5-alacritty-unstable-2018-08-05/bin
patching script interpreter paths in /nix/store/04d6b7v15l2mg7aq24xsf9xzgjl18rj5-alacritty-unstable-2018-08-05
checking for references to /build in /nix/store/04d6b7v15l2mg7aq24xsf9xzgjl18rj5-alacritty-unstable-2018-08-05...
strip is /nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/hgn9kc9kixyfqm4wk5vqv4m842xlqysi-alacritty-unstable-2018-08-05-terminfo
checking for references to /build in /nix/store/hgn9kc9kixyfqm4wk5vqv4m842xlqysi-alacritty-unstable-2018-08-05-terminfo...
/nix/store/04d6b7v15l2mg7aq24xsf9xzgjl18rj5-alacritty-unstable-2018-08-05
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=231b6a66-b365-4844-a585-2460a4f246af-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46362&attempt_id=7f5dee7a-1498-48dc-afc7-976cbed9734f)

Attempted: fd

<details><summary>Partial log (click to expand)</summary><p>

```
test result: ok. 36 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

installing
post-installation fixup
moving /nix/store/vk31ac3j7in9dkyi0znzgjc9wgf30vsq-fd-7.1.0/man to /nix/store/vk31ac3j7in9dkyi0znzgjc9wgf30vsq-fd-7.1.0/share/man
gzipping man pages under /nix/store/vk31ac3j7in9dkyi0znzgjc9wgf30vsq-fd-7.1.0/share/man/
strip is /nix/store/df6k4mgdjxciy0f637lryp7c9ln7n1m3-cctools-binutils-darwin/bin/strip
stripping (with command strip and flags -S) in /nix/store/vk31ac3j7in9dkyi0znzgjc9wgf30vsq-fd-7.1.0/bin
patching script interpreter paths in /nix/store/vk31ac3j7in9dkyi0znzgjc9wgf30vsq-fd-7.1.0
/nix/store/vk31ac3j7in9dkyi0znzgjc9wgf30vsq-fd-7.1.0
```
</p></details>

Mic92 commented:
Just if alacritty can vendor its source. If alacritty itself builds is a different story.

GrahamcOfBorg commented:
<!--REQUEST_ID=231b6a66-b365-4844-a585-2460a4f246af-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46362&attempt_id=81f5165c-aea3-4253-9a17-ac0ff15ad35a)

Attempted: fd

<details><summary>Partial log (click to expand)</summary><p>

```
post-installation fixup
moving /nix/store/5314a1w9s9vqg5b4w9205qi8j0ywzh7j-fd-7.1.0/man to /nix/store/5314a1w9s9vqg5b4w9205qi8j0ywzh7j-fd-7.1.0/share/man
shrinking RPATHs of ELF executables and libraries in /nix/store/5314a1w9s9vqg5b4w9205qi8j0ywzh7j-fd-7.1.0
shrinking /nix/store/5314a1w9s9vqg5b4w9205qi8j0ywzh7j-fd-7.1.0/bin/fd
gzipping man pages under /nix/store/5314a1w9s9vqg5b4w9205qi8j0ywzh7j-fd-7.1.0/share/man/
strip is /nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/5314a1w9s9vqg5b4w9205qi8j0ywzh7j-fd-7.1.0/bin
patching script interpreter paths in /nix/store/5314a1w9s9vqg5b4w9205qi8j0ywzh7j-fd-7.1.0
checking for references to /build in /nix/store/5314a1w9s9vqg5b4w9205qi8j0ywzh7j-fd-7.1.0...
/nix/store/5314a1w9s9vqg5b4w9205qi8j0ywzh7j-fd-7.1.0
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=231b6a66-b365-4844-a585-2460a4f246af-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46362&attempt_id=4a198c4b-fecb-4469-8e39-e9e3eee9852f)

Attempted: fd

<details><summary>Partial log (click to expand)</summary><p>

```
post-installation fixup
moving /nix/store/0rinyr3nwsl0iy32v8nandj6w9gkbn6k-fd-7.1.0/man to /nix/store/0rinyr3nwsl0iy32v8nandj6w9gkbn6k-fd-7.1.0/share/man
shrinking RPATHs of ELF executables and libraries in /nix/store/0rinyr3nwsl0iy32v8nandj6w9gkbn6k-fd-7.1.0
shrinking /nix/store/0rinyr3nwsl0iy32v8nandj6w9gkbn6k-fd-7.1.0/bin/fd
gzipping man pages under /nix/store/0rinyr3nwsl0iy32v8nandj6w9gkbn6k-fd-7.1.0/share/man/
strip is /nix/store/y4ymnvgxygpq05h03kyzbj572zmh6zla-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/0rinyr3nwsl0iy32v8nandj6w9gkbn6k-fd-7.1.0/bin
patching script interpreter paths in /nix/store/0rinyr3nwsl0iy32v8nandj6w9gkbn6k-fd-7.1.0
checking for references to /build in /nix/store/0rinyr3nwsl0iy32v8nandj6w9gkbn6k-fd-7.1.0...
/nix/store/0rinyr3nwsl0iy32v8nandj6w9gkbn6k-fd-7.1.0
```
</p></details>

symphorien committed:
cargo-vendor-normalise: add a small install check

symphorien commented:
Rebased.
I also have added a small check that cargo-vendor-normalise outputs a style consistent with the committed template.

Mic92 commented:
@GrahamcOfBorg build fd

Mic92 commented:
@GrahamcOfBorg build alacritty

Mic92 commented:
too bad, staging will never finish in time.

GrahamcOfBorg commented:
<!--REQUEST_ID=f2d11a7c-371e-4abf-aa03-f11ad4ab0249-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46362&attempt_id=94eb011f-84ff-485c-aff8-69c1c39da94d)

Attempted: fd

<details><summary>Partial log (click to expand)</summary><p>

```
post-installation fixup
moving /nix/store/f6k43h2zp448skjf1252hgndwhlsrz54-fd-7.1.0/man to /nix/store/f6k43h2zp448skjf1252hgndwhlsrz54-fd-7.1.0/share/man
shrinking RPATHs of ELF executables and libraries in /nix/store/f6k43h2zp448skjf1252hgndwhlsrz54-fd-7.1.0
shrinking /nix/store/f6k43h2zp448skjf1252hgndwhlsrz54-fd-7.1.0/bin/fd
gzipping man pages under /nix/store/f6k43h2zp448skjf1252hgndwhlsrz54-fd-7.1.0/share/man/
strip is /nix/store/rf59n87kcrrnrnw3l3s94q879m8vaspm-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/f6k43h2zp448skjf1252hgndwhlsrz54-fd-7.1.0/bin
patching script interpreter paths in /nix/store/f6k43h2zp448skjf1252hgndwhlsrz54-fd-7.1.0
checking for references to /build in /nix/store/f6k43h2zp448skjf1252hgndwhlsrz54-fd-7.1.0...
/nix/store/f6k43h2zp448skjf1252hgndwhlsrz54-fd-7.1.0
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=f2d11a7c-371e-4abf-aa03-f11ad4ab0249-->
Timed out, unknown build status on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46362&attempt_id=f4fe5c3f-bc7d-4436-85ac-73ab6cc20f45)

Attempted: fd

<details><summary>Partial log (click to expand)</summary><p>

```
cannot build derivation '/nix/store/ajp1dq347yjpiw53i59cizdd0dw43q90-apple-framework-Security.drv': 5 dependencies couldn't be built
cannot build derivation '/nix/store/mmspxib9ar4d1mld1zn29hiwlswy6l56-docbook2X-0.8.8.drv': 16 dependencies couldn't be built
cannot build derivation '/nix/store/m14n58jg1mcqg3ywiwz98dgdvi2y7acj-git-2.18.0.drv': 29 dependencies couldn't be built
cannot build derivation '/nix/store/dv0znvjk4syz4255jw06k73sfzbjv9i5-libgit2-0.26.6.drv': 12 dependencies couldn't be built
cannot build derivation '/nix/store/q6srb8vb4bdvi0swwxk73dhi8zs2vwpf-rustc-bootstrap-1.26.2.drv': 6 dependencies couldn't be built
cannot build derivation '/nix/store/l9v5l4fmvzw1d74hax8ysn64cv0cv8ii-cargo-bootstrap-1.26.2.drv': 8 dependencies couldn't be built
cannot build derivation '/nix/store/mi6z331lvzd2iqy8vw88161l3a911r2i-rustc-1.27.0.drv': 15 dependencies couldn't be built
cannot build derivation '/nix/store/g88y3qgqnd0igmmynyj03k4jlbgw8qv1-cargo-1.27.0.drv': 18 dependencies couldn't be built
cannot build derivation '/nix/store/0c5c9wn2809krxcdcjkmaq0rpd9xswlr-fd-7.1.0.drv': 6 dependencies couldn't be built
error: build of '/nix/store/0c5c9wn2809krxcdcjkmaq0rpd9xswlr-fd-7.1.0.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=f2d11a7c-371e-4abf-aa03-f11ad4ab0249-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46362&attempt_id=62d7b3b3-55f0-4a74-b4f0-898100053dc0)

Attempted: fd

<details><summary>Partial log (click to expand)</summary><p>

```
(cd perl/build/lib && tar cf - .) | \
(cd '/nix/store/slqgvaj54rls5sgsvsq6q5040zkn8fma-git-2.18.0/lib/perl5/site_perl/5.28.0' && umask 022 && tar xof -)
tar: Skipping to next header
tar: Exiting with failure status due to previous errors
make: *** [Makefile:2733: install] Error 2
builder for '/nix/store/819b6207zpq9zigp3rii0lrvvngicl84-git-2.18.0.drv' failed with exit code 2
cannot build derivation '/nix/store/7vgqhdlcx5sgic3w1d7a7si4wz8k1m6q-rustc-1.27.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/3y74nvymh7b0n0vbqcdcsjna4h2yd5jx-cargo-1.27.0.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/jwx0h9p1h7biaci6dj3vmb2h9p0rvhiq-fd-7.1.0.drv': 3 dependencies couldn't be built
error: build of '/nix/store/jwx0h9p1h7biaci6dj3vmb2h9p0rvhiq-fd-7.1.0.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=43d80b71-b7f6-4103-bb65-b152467d6580-->
Timed out, unknown build status on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46362&attempt_id=13d4a10d-d321-4146-80e1-d297e859b8fb)

Attempted: alacritty

<details><summary>Partial log (click to expand)</summary><p>

```
cannot build derivation '/nix/store/0fg2mszl1rmld1avjpxndwkqw0jpff11-rust_ignore-0.2.2.drv': 21 dependencies couldn't be built
cannot build derivation '/nix/store/f8yqibkm8i9sy0qfkddg1ixxylws6c1p-apple-framework-AudioToolbox.drv': 7 dependencies couldn't be built
cannot build derivation '/nix/store/jc0gc3frjqfxq61syvqk1lvpb6s1frd8-mesa-noglu-18.1.8.drv': 32 dependencies couldn't be built
cannot build derivation '/nix/store/yjxpivc3qr37ba1ysmb6z0c4rjrj2phv-rust_cargo-0.22.0.drv': 86 dependencies couldn't be built
cannot build derivation '/nix/store/p25bnbmhmscxnwhh7r7qrl60jkyx9nwy-apple-framework-AppKit.drv': 6 dependencies couldn't be built
cannot build derivation '/nix/store/whdi2yfjx5hkhs8i0c911b6x5lcch64d-libGL-1.0.0.drv': 5 dependencies couldn't be built
cannot build derivation '/nix/store/213n3kij0pm6qam2n7j11m446pzyi2hw-rust_cargo-vendor-0.1.13.drv': 85 dependencies couldn't be built
cannot build derivation '/nix/store/g4fwpg2qyyk2xxfpxcdpnjxjrhccbm0j-alacritty-unstable-2018-08-05-vendor.drv': 7 dependencies couldn't be built
cannot build derivation '/nix/store/cxbxcizjra3vb1pd53az4l6vlq7xpb6z-alacritty-unstable-2018-08-05.drv': 29 dependencies couldn't be built
error: build of '/nix/store/cxbxcizjra3vb1pd53az4l6vlq7xpb6z-alacritty-unstable-2018-08-05.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=43d80b71-b7f6-4103-bb65-b152467d6580-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46362&attempt_id=5c87fe37-f1a2-4c31-9487-cc73c00d4ff0)

Attempted: alacritty

<details><summary>Partial log (click to expand)</summary><p>

```
checking for references to /build in /nix/store/amignn5wl2ljwhl953wdr8ladjwdl6mw-mesa-noglu-18.1.8-drivers...
shrinking RPATHs of ELF executables and libraries in /nix/store/zdagshlqwn5idnjdmq9gfyjbyaldza8j-mesa-noglu-18.1.8-osmesa
shrinking /nix/store/zdagshlqwn5idnjdmq9gfyjbyaldza8j-mesa-noglu-18.1.8-osmesa/lib/libOSMesa.so.8.0.0
strip is /nix/store/lw308kxqcm9frx4vlfjh30da1x4gsmwx-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/zdagshlqwn5idnjdmq9gfyjbyaldza8j-mesa-noglu-18.1.8-osmesa/lib
patching script interpreter paths in /nix/store/zdagshlqwn5idnjdmq9gfyjbyaldza8j-mesa-noglu-18.1.8-osmesa
checking for references to /build in /nix/store/zdagshlqwn5idnjdmq9gfyjbyaldza8j-mesa-noglu-18.1.8-osmesa...
building '/nix/store/d1mr1nzmlfwzq8rpmg242hffhyx3141f-libGL-1.0.0.drv'...
cannot build derivation '/nix/store/ls0n3pgsj5pifsq9askwghyd457z52k1-alacritty-unstable-2018-08-05.drv': 3 dependencies couldn't be built
error: build of '/nix/store/ls0n3pgsj5pifsq9askwghyd457z52k1-alacritty-unstable-2018-08-05.drv' failed
```
</p></details>

Mic92 commented:
@LnL7 ok, however could you just check if the cargo vendor normalization works for you? ofBorg returned an failure on macOS. Just take any rust package and change the `cargoSha256` checksum.

Mic92 commented on pkgs/build-support/rust/fetchcargo.nix:
```diff
@@ -1,8 +1,24 @@
+    < $reference $out/bin/cargo-vendor-normalise > test;
+    cmp test $reference
+  '';
+  buildInputs = [ (python3.withPackages(ps: [ ps.toml ])) ];
```
I just realized we cannot use the results `python3.withPackages` to replace the shebang of the shell script on macOS, instead we must call the wrapped python executable and pass `cargo-vendor-normalise.py` as the first argument. 
On macOS shebangs must be binaries and cannot be scripts with shebangs itself.

Mic92 replied:
this is what @Infinisil got on his machine: https://hastebin.com/uxihohicax.sql

infinisil committed:
fetchcargo: Fix cargo-vendor-normalise for darwin

Infinisil commented:
Added a commit that fixes this.

Trying an Alacritty build on Darwin with a slightly changed dependency hash leads to:

```
To use vendored sources, add this to your .cargo/config for this project:
output path ‘/nix/store/0wqzscwlczmb8s61jc66wwqjpv9milfr-alacritty-unstable-2018-08-05-vendor’ has r:sha256 hash
‘0ms0248bb2lgbzcqks6i0qhn1gaiim3jf1kl17qw52c8an3rc652’ when 
‘1ms0248bb2lgbzcqks6i0qhn1gaiim3jf1kl17qw52c8an3rc652’ was expected
cannot build derivation ‘/nix/store/jkk6kj7vbn2ir9z6rj29liivdzx3zdxy-alacritty-unstable-2018-08-05.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/jkk6kj7vbn2ir9z6rj29liivdzx3zdxy-alacritty-unstable-2018-08-05.drv’ failed
```

So the hashes are indeed the same :)

I cherry-picked these commits to master so I wouldn't have to build all of stdenv ([branch](https://github.com/Infinisil/nixpkgs/commits/cargo))

Infinisil commented:
Alacritty fully builds on the cherry-picked master branch even :)

Mic92 commented:
@Infinisil thanks!

symphorien commented:
So this is the "right way" to wrap python scripts...
Thanks!

akru commented:
@Mic92 https://github.com/NixOS/nixpkgs/commit/dc8aca448d0570ade231f92272c4453bc9837fc9 broke this.

Config looked at https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/rust/default.nix#L60 by name `config` but https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/rust/fetchcargo.nix#L53 install it with name `cargo-config.XXX`.
